flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveScheduler.java

```diff
@@ -911,19 +914,26 @@ public void runIfState(State expectedState, Runnable action, Duration delay) {
 
     // ----------------------------------------------------------------
 
+    /** Note: Do not call this method from a State constructor. */
     @VisibleForTesting
-    void transitionToState(State newState) {
-        if (state != newState) {
+    void transitionToState(StateFactory<?> targetState) {
+        Preconditions.checkState(
+                !isTransitioningState, "Concurrent state transitions are not allowed");
```

Comment:
I'm wondering if "concurrent" is the right word; to me it usually implies threading issues and such.

How about "State transitions must not be triggered while another state transition is in progress."